pkg/storage/allocator.go

```diff
@@ -491,6 +511,33 @@ func (a Allocator) RebalanceTarget(
 	if target == nil {
 		return nil, ""
 	}
+	// We could make a simulation here to verify whether we'll remove the target we'll rebalance to.
+	if repl.RaftStatus() != nil {
+		newReplica := roachpb.ReplicaDescriptor{
+			NodeID:    target.store.Node.NodeID,
+			StoreID:   target.store.StoreID,
+			ReplicaID: rangeInfo.Desc.NextReplicaID,
+		}
+		desc := *rangeInfo.Desc
+		desc.Replicas = append(desc.Replicas, newReplica)
+		rangeInfo.Desc = &desc
+
+		replicaCandidates := simulationFilterUnremovableReplicas(repl, desc.Replicas, newReplica.ReplicaID)
+
+		removeReplica, _, err := a.simulationRemoveTarget(ctx, repl, target.store.StoreID, constraints, replicaCandidates, rangeInfo)
+		if err != nil {
+			log.Warningf(ctx, "simulating RemoveTarget failed: %s", err)
+			return nil, ""
+		}
+		if removeReplica.StoreID == target.store.StoreID {
+			newTargets := candidates.removeCandidate(*target)
+			newTarget := newTargets.selectGood(a.randGen)
+			if newTarget == nil {
+				return nil, ""
+			}
+			target = newTarget
```

Comment:
I think we actually want this logic to happen in a loop. What if the `newTarget` is also going to be immediately removed?

Reply:
Yeah, You are right. My mistake, Done.